Improve BasicBlock dumping

BruceForstall opened the pull request:
1. Display both true and false target destinations of `BBJ_COND` blocks, e.g. `BB229,BB226 ( cond )`. The first block listed is the true target; the second block is the false target.
2. If a branch goes to the next block that will be displayed in the table, display "*" instead of the block number, as a shorthand to reduce visual clutter, e.g., `BB14,*      ( cond )`

BruceForstall committed:
Improve BasicBlock dumping

1. Display both true and false target destinations of blocks
2. If a branch goes to the next block that will be displayed
in the table, display "*" instead of the block number, as a
shorthand to reduce visual clutter.

BruceForstall commented:
Sample block table dump:
```
----------------------------------------------------------------------------------------------------------------------------------------------
BBnum  BBid ref try hnd preds           weight     IBC [IL range]   [jump]                [EH region]        [flags]
----------------------------------------------------------------------------------------------------------------------------------------------
BB01  [0795]  1                             1       11 [???..???)-> *           (always)                     i LIR IBC internal label q
BB02  [0000]  1       BB01                  1       11 [000..015)-> BB68,*      ( cond )                     i LIR IBC
BB03  [0001]  1       BB02                  0.27     3 [015..016)-> BB277,*     ( cond )                     i LIR IBC
BB04  [0812]  2       BB03,BB278            1.09    12 [015..016)-> *           (always)                     i LIR IBC label bwd preheader q
BB06  [0049]  2       BB04,BB57            34.91   384 [015..016)-> BB38,*      ( cond )                     i LIR IBC label bwd bwd-target
BB07  [0050]  1       BB06                 18.44   203 [015..016)-> BB08,*      ( cond )                     i LIR IBC bwd
BB266 [0752] 37       BB07,BB09,BB10,BB14,BB15,BB23,BB28,BB30,BB38,BB39,BB48,BB50,BB62,BB74,BB85,BB96,BB107,BB116,BB152,BB154,BB155,BB159,BB160,BB168,BB173,BB184,BB185,BB187,BB190,BB204,BB218,BB229,BB237,BB238,BB247,BB258,BB259   0        0 [19B..19C)               (throw )                     i LIR IBC rare label hascall gcsafe bwd q
BB277 [0817]  1       BB03                  0.14     2 [???..???)-> BB59        (always)                     LIR IBC internal label has-align
BB08  [0059]  1       BB07                 18.44   203 [015..016)-> BB14,*      ( cond )                     i LIR IBC label hascall gcsafe bwd
BB09  [0065]  1       BB08                 18.27   201 [015..016)-> BB266,*     ( cond )                     i LIR IBC bwd
BB10  [0071]  1       BB09                 18.27   201 [015..016)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB11  [0078]  1       BB10                 18.27   201 [015..016)-> BB13,*      ( cond )                     i LIR IBC hascall gcsafe bwd
BB12  [0082]  1       BB11                 18.19   200 [015..016)-> BB19        (always)                     i LIR IBC hascall gcsafe bwd
BB13  [0083]  1       BB11                 18.27   201 [015..016)-> *           (always)                     i LIR IBC label hascall gcsafe bwd q
BB19  [0103]  4       BB12,BB13,BB17,BB18  18.44   203 [015..016)-> BB23,*      ( cond )                     i LIR IBC label bwd
BB20  [0108]  1       BB19                 18.43   203 [015..016)-> BB21,*      ( cond )                     i LIR IBC bwd
BB22  [0110]  1       BB20                  0.01     0 [015..016)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB23  [0111]  3       BB19,BB21,BB22       18.44   203 [015..016)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB24  [0115]  1       BB23                 18.44   203 [015..016)-> *           (always)                     i LIR IBC hascall gcsafe bwd preheader q
BB26  [0122]  2       BB24,BB25            26.21   288 [015..016)-> BB28,*      ( cond )                     i LIR IBC label bwd align
BB27  [0123]  1       BB26                 26.61   293 [015..016)-> BB25,*      ( cond )                     i LIR IBC bwd bwd-src
BB28  [0124]  2       BB26,BB27            18.44   203 [015..016)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB29  [0128]  1       BB28                 18.44   203 [015..016)-> BB30,*      ( cond )                     i LIR IBC hascall gcsafe bwd
BB36  [0133]  1       BB29                  0        0 [015..016)-> *           (always)                     i LIR IBC rare bwd q
BB37  [0134]  2       BB35,BB36            18.44   203 [015..016)-> *           (always)                     i LIR IBC label bwd q
BB38  [0051]  2       BB06,BB37            34.91   384 [015..016)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB39  [0151]  1       BB38                 34.91   384 [015..016)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB40  [0158]  1       BB39                 34.91   384 [015..016)-> BB44,*      ( cond )                     i LIR IBC hascall gcsafe bwd
BB41  [0168]  1       BB40                 34.90   384 [015..016)-> BB42,*      ( cond )                     i LIR IBC bwd
BB43  [0170]  1       BB41                  0.01     0 [015..016)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB44  [0171]  3       BB40,BB42,BB43       34.91   384 [015..016)-> *           (always)                     i LIR IBC label gcsafe bwd preheader has-align q
BB46  [0182]  2       BB44,BB45            49.63   546 [015..016)-> BB48,*      ( cond )                     i LIR IBC label bwd align
BB47  [0183]  1       BB46                 50.39   554 [015..016)-> BB45,*      ( cond )                     i LIR IBC bwd bwd-src
BB48  [0184]  2       BB46,BB47            34.91   384 [015..016)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB49  [0188]  1       BB48                 34.91   384 [015..016)-> BB50,*      ( cond )                     i LIR IBC hascall gcsafe bwd
BB56  [0193]  1       BB49                  0        0 [015..016)-> *           (always)                     i LIR IBC rare bwd q
BB57  [0194]  2       BB55,BB56            34.91   384 [015..016)-> BB06,*      ( cond )                     i LIR IBC label bwd
BB58  [0053]  1       BB57                  1.09    12 [015..016)-> BB278,*     ( cond )                     i LIR IBC bwd
BB59  [0055]  2       BB58,BB277            0.27     3 [015..02D)-> BB60,*      ( cond )                     i LIR IBC label hascall gcsafe
BB61  [0210]  1       BB59                  0        0 [024..025)               (throw )                     i LIR IBC rare hascall gcsafe q
BB278 [0818]  1       BB58                  0.55     6 [???..???)-> BB04        (always)                     LIR IBC internal label bwd
BB60  [0209]  1       BB59                  0.27     3 [024..025)-> BB62        (always)                     i LIR IBC label hascall gcsafe has-align q
BB50  [0192]  1       BB49                 34.91   384 [015..016)-> BB266,*     ( cond )                     i LIR IBC label hascall gcsafe bwd
BB51  [0197]  1       BB50                 34.91   384 [015..016)-> *           (always)                     i LIR IBC hascall gcsafe bwd preheader q
BB53  [0204]  2       BB51,BB52            49.63   546 [015..016)-> BB55,*      ( cond )                     i LIR IBC label bwd align
BB54  [0205]  1       BB53                 50.39   554 [015..016)-> BB52,*      ( cond )                     i LIR IBC bwd bwd-src
BB55  [0206]  2       BB53,BB54            34.91   384 [015..016)-> BB57        (always)                     i LIR IBC label bwd q
BB183 [0557]  1       BB181                 0.02     0 [19B..19C)-> BB184       (always)                     i LIR IBC label hascall gcsafe bwd has-align q
BB42  [0169]  1       BB41                 34.89   384 [015..016)-> BB44        (always)                     i LIR IBC label bwd
BB179 [0543]  1       BB177                 0.02     0 [19B..19C)-> BB180       (always)                     i LIR IBC label hascall gcsafe bwd q
BB30  [0132]  1       BB29                 18.44   203 [015..016)-> BB266,*     ( cond )                     i LIR IBC label hascall gcsafe bwd
BB31  [0137]  1       BB30                 18.44   203 [015..016)-> *           (always)                     i LIR IBC hascall gcsafe bwd preheader q
BB33  [0144]  2       BB31,BB32            26.21   288 [015..016)-> BB35,*      ( cond )                     i LIR IBC label bwd align
BB34  [0145]  1       BB33                 26.61   293 [015..016)-> BB32,*      ( cond )                     i LIR IBC bwd bwd-src
BB35  [0146]  2       BB33,BB34            18.44   203 [015..016)-> BB37        (always)                     i LIR IBC label bwd q
BB176 [0536]  1       BB174                61.27   674 [19B..19C)-> BB180,BB177 ( cond )                     i LIR IBC label bwd
BB21  [0109]  1       BB20                 18.43   203 [015..016)-> BB23        (always)                     i LIR IBC label bwd
BB177 [0541]  1       BB176                61.26   674 [19B..19C)-> BB179,BB178 ( cond )                     i LIR IBC label bwd
BB14  [0066]  1       BB08                  0.17     2 [015..016)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB15  [0087]  1       BB14                  0.17     2 [015..016)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB16  [0094]  1       BB15                  0.17     2 [015..016)-> BB17,*      ( cond )                     i LIR IBC hascall gcsafe bwd
BB18  [0099]  1       BB16                  0.00     0 [015..016)-> BB19        (always)                     i LIR IBC hascall gcsafe bwd has-align q
BB178 [0542]  1       BB177                61.24   674 [19B..19C)-> BB180       (always)                     i LIR IBC label bwd q
BB17  [0098]  1       BB16                  0.17     2 [015..016)-> BB19        (always)                     i LIR IBC label hascall gcsafe bwd
BB180 [0544]  3       BB176,BB178,BB179    61.27   674 [19B..19C)-> BB184,BB181 ( cond )                     i LIR IBC label hascall gcsafe bwd
BB25  [0121]  1       BB27                  7.78    86 [015..016)-> BB26        (always)                     i LIR IBC label bwd bwd-target q
BB181 [0555]  1       BB180                61.26   674 [19B..19C)-> BB183,BB182 ( cond )                     i LIR IBC label bwd
BB32  [0143]  1       BB34                  7.78    86 [015..016)-> BB33        (always)                     i LIR IBC label bwd bwd-target q
BB182 [0556]  1       BB181                61.24   674 [19B..19C)-> BB184       (always)                     i LIR IBC label bwd q
BB45  [0181]  1       BB47                 14.72   162 [015..016)-> BB46        (always)                     i LIR IBC label bwd bwd-target q
BB184 [0558]  3       BB180,BB182,BB183    61.27   674 [19B..19C)-> BB266,BB185 ( cond )                     i LIR IBC label bwd
BB52  [0203]  1       BB54                 14.72   162 [015..016)-> BB53        (always)                     i LIR IBC label bwd bwd-target q
BB185 [0562]  1       BB184                61.27   674 [19B..19C)-> BB266,BB186 ( cond )                     i LIR IBC label hascall gcsafe bwd
BB62  [0211]  1       BB60                  0.27     3 [02D..02E)-> BB266,*     ( cond )                     i LIR IBC label
BB63  [0219]  1       BB62                  0.27     3 [02D..043)-> BB67,*      ( cond )                     i LIR IBC hascall gcsafe
BB64  [0229]  1       BB63                  0.27     3 [03C..03D)-> BB65,*      ( cond )                     i LIR IBC
BB66  [0231]  1       BB64                  0.00     0 [03C..03D)-> *           (always)                     i LIR IBC hascall gcsafe q
BB67  [0232]  6       BB63,BB65,BB66,BB272,BB274,BB275   0.27     3 [03D..044)               (return)                     i LIR IBC label
BB186 [0569]  1       BB185                61.27   674 [19B..19C)-> BB189,*     ( cond )                     i LIR IBC label hascall gcsafe bwd
BB187 [0575]  1       BB186                61.06   672 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC bwd
BB188 [0585]  1       BB187                61.06   672 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB189 [0576]  2       BB186,BB188          61.27   674 [19B..19C)-> BB192,*     ( cond )                     i LIR IBC label bwd
BB190 [0577]  1       BB189                61.06   672 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC bwd
BB191 [0590]  1       BB190                61.06   672 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB192 [0578]  2       BB189,BB191          61.27   674 [19B..19C)-> BB195,*     ( cond )                     i LIR IBC label bwd
BB193 [0807]  2       BB192,BB193          3025. 33273 [19B..19C)-> BB193,*     ( cond )                     i LIR IBC label bwd align has-align
BB195 [0596]  3       BB192,BB193,BB197   145.45  1600 [19B..19C)-> BB279,*     ( cond )                     i LIR IBC label bwd align
BB199 [0599]  1       BB195                 0        0 [19B..19C)-> BB197,*     ( cond )                     i LIR IBC rare bwd bwd-src
BB280 [0820]  1       BB199                 0        0 [???..???)-> *           (always)                     LIR IBC rare internal bwd q
BB200 [0600]  2       BB279,BB280          61.27   674 [19B..19C)-> *           (always)                     i LIR IBC label hascall gcsafe bwd preheader has-align q
BB202 [0607]  2       BB200,BB201          87.12   958 [19B..19C)-> BB204,*     ( cond )                     i LIR IBC label bwd align
BB203 [0608]  1       BB202                88.44   973 [19B..19C)-> BB201,*     ( cond )                     i LIR IBC bwd bwd-src
BB204 [0609]  2       BB202,BB203          61.27   674 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB205 [0613]  1       BB204                61.27   674 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd preheader q
BB220 [0580]  2       BB205,BB219          64.02   704 [19B..19C)-> BB221,*     ( cond )                     i LIR IBC label hascall gcsafe bwd bwd-src
BB229 [0642]  2       BB220,BB224           0        0 [19B..19C)-> BB266,BB231 ( cond )                     i LIR IBC rare label bwd
BB279 [0819]  1       BB195                72.73   800 [???..???)-> BB200       (always)                     LIR IBC internal label bwd has-align
BB221 [0643]  1       BB220                64.02   704 [19B..19C)-> BB206,*     ( cond )                     i LIR IBC label bwd
BB223 [0645]  1       BB221                64.02   704 [19B..19C)-> BB206,*     ( cond )                     i LIR IBC bwd
BB224 [0809]  2       BB223,BB227          64.02   704 [19B..19C)-> BB229,BB226 ( cond )                     i LIR IBC label bwd align
BB201 [0606]  1       BB203                25.84   284 [19B..19C)-> BB202       (always)                     i LIR IBC label bwd bwd-target q
BB206 [0579]  4       BB221,BB223,BB226,BB227   2.75    30 [19B..19C)-> BB209,*     ( cond )                     i LIR IBC label bwd bwd-target has-align
BB207 [0810]  2       BB206,BB207         135.60  1492 [19B..19C)-> BB207,*     ( cond )                     i LIR IBC label bwd align has-align
BB209 [0619]  3       BB206,BB207,BB211   581.82  6400 [19B..19C)-> BB214,*     ( cond )                     i LIR IBC label bwd align
BB213 [0622]  1       BB209                 0        0 [19B..19C)-> BB211,*     ( cond )                     i LIR IBC rare bwd bwd-src
BB214 [0623]  2       BB209,BB213           2.75    30 [19B..19C)-> *           (always)                     i LIR IBC label hascall gcsafe bwd preheader has-align q
BB216 [0630]  2       BB214,BB215           3.91    43 [19B..19C)-> BB218,*     ( cond )                     i LIR IBC label bwd align
BB217 [0631]  1       BB216                 3.96    44 [19B..19C)-> BB215,*     ( cond )                     i LIR IBC bwd bwd-src
BB218 [0632]  2       BB216,BB217           2.75    30 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB219 [0636]  1       BB218                 2.75    30 [19B..19C)-> BB220       (always)                     i LIR IBC hascall gcsafe bwd has-align
BB244 [0704]  1       BB246                50.30   553 [19B..19C)-> BB245       (always)                     i LIR IBC label bwd bwd-target align q
BB215 [0629]  1       BB217                 1.16    13 [19B..19C)-> BB216       (always)                     i LIR IBC label bwd bwd-target q
BB242 [0693]  1       BB240                 0.04     0 [19B..19C)-> BB243       (always)                     i LIR IBC label hascall gcsafe bwd q
BB226 [0648]  1       BB224                 2.92    32 [19B..19C)-> BB206,*     ( cond )                     i LIR IBC label bwd
BB227 [0650]  1       BB226                64.02   704 [19B..19C)-> BB206,BB224 ( cond )                     i LIR IBC bwd
BB236 [0533]  2       BB175,BB235          61.27   674 [19B..19C)-> *           (always)                     i LIR IBC label bwd q
BB237 [0450]  2       BB151,BB236         119.27  1312 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB238 [0674]  1       BB237               119.27  1312 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB239 [0681]  1       BB238               119.27  1312 [19B..19C)-> BB243,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB240 [0691]  1       BB239               119.25  1312 [19B..19C)-> BB242,*     ( cond )                     i LIR IBC bwd
BB241 [0692]  1       BB240               119.21  1311 [19B..19C)-> BB243       (always)                     i LIR IBC bwd
BB65  [0230]  1       BB64                  0.27     3 [03C..03D)-> BB67        (always)                     i LIR IBC label
BB243 [0694]  3       BB239,BB241,BB242   119.27  1312 [19B..19C)-> *           (always)                     i LIR IBC label gcsafe bwd preheader q
BB245 [0705]  2       BB243,BB244         169.58  1865 [19B..19C)-> BB247,*     ( cond )                     i LIR IBC label bwd
BB246 [0706]  1       BB245               172.15  1894 [19B..19C)-> BB244,*     ( cond )                     i LIR IBC bwd bwd-src
BB247 [0707]  2       BB245,BB246         119.27  1312 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB248 [0711]  1       BB247               119.27  1312 [19B..19C)-> BB250,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB249 [0715]  1       BB248                 9.09       [19B..19C)-> BB261       (always)                     i LIR bwd
BB68  [0002]  1       BB02                  0.73     8 [044..057)-> BB70,*      ( cond )                     i LIR IBC label
BB69  [0003]  1       BB68                  0        0 [000..06C)-> BB74        (always)                     i LIR IBC rare has-align q
BB70  [0004]  1       BB68                  0.73     8 [000..000)-> BB72,*      ( cond )                     i LIR IBC label gcsafe
BB71  [0244]  1       BB70                  0.32     4 [000..000)-> BB73        (always)                     i LIR IBC internal
BB72  [0245]  1       BB70                  0.41     4 [000..000)-> *           (always)                     i LIR IBC internal label idxlen nullcheck q
BB73  [0249]  2       BB71,BB72             0.73     8 [080..080)-> *           (always)                     i LIR IBC internal label q
BB74  [0005]  2       BB69,BB73             0.73     8 [080..081)-> BB266,*     ( cond )                     i LIR IBC label
BB75  [0252]  1       BB74                  0.73     8 [080..092)-> BB79,*      ( cond )                     i LIR IBC hascall gcsafe
BB76  [0262]  1       BB75                  0.73     8 [08B..08C)-> BB77,*      ( cond )                     i LIR IBC
BB78  [0264]  1       BB76                  0.00     0 [08B..08C)-> *           (always)                     i LIR IBC hascall gcsafe q
BB79  [0265]  3       BB75,BB77,BB78        0.73     8 [08C..0AC)-> BB81,*      ( cond )                     i LIR IBC label
BB80  [0006]  1       BB79                  0.18     2 [000..0C1)-> BB85        (always)                     i LIR IBC q
BB77  [0263]  1       BB76                  0.73     8 [08B..08C)-> BB79        (always)                     i LIR IBC label
BB148 [0035]  5       BB117,BB139,BB140,BB141,BB147   0.73     8 [19B..19C)-> *,BB149     ( cond )                     i LIR IBC label
BB281 [0821]  1       BB148                 0.36     4 [???..???)-> BB263       (always)                     LIR IBC internal label
BB81  [0007]  1       BB79                  0.55     6 [000..000)-> BB83,*      ( cond )                     i LIR IBC label gcsafe
BB82  [0277]  1       BB81                  0.24     3 [000..000)-> BB84        (always)                     i LIR IBC internal
BB83  [0278]  1       BB81                  0.31     3 [000..000)-> *           (always)                     i LIR IBC internal label idxlen nullcheck q
BB84  [0282]  2       BB82,BB83             0.55     6 [0D6..0D6)-> *           (always)                     i LIR IBC internal label q
BB85  [0008]  2       BB80,BB84             0.73     8 [0D6..0D7)-> BB266,*     ( cond )                     i LIR IBC label
BB86  [0285]  1       BB85                  0.73     8 [0D6..0E8)-> BB90,*      ( cond )                     i LIR IBC hascall gcsafe
BB87  [0295]  1       BB86                  0.73     8 [0E1..0E2)-> BB88,*      ( cond )                     i LIR IBC
BB89  [0297]  1       BB87                  0.00     0 [0E1..0E2)-> *           (always)                     i LIR IBC hascall gcsafe q
BB90  [0298]  3       BB86,BB88,BB89        0.73     8 [0E2..0FC)-> BB92,*      ( cond )                     i LIR IBC label
BB91  [0009]  1       BB90                  0        0 [000..111)-> BB96        (always)                     i LIR IBC rare q
BB88  [0296]  1       BB87                  0.73     8 [0E1..0E2)-> BB90        (always)                     i LIR IBC label
BB149 [0805]  2       BB148,BB282           3.73    41 [19B..19C)-> *           (always)                     i LIR IBC label bwd preheader q
BB151 [0448]  2       BB149,BB261         119.27  1312 [19B..19C)-> BB237,BB152 ( cond )                     i LIR IBC label bwd bwd-target
BB92  [0010]  1       BB90                  0.73     8 [000..000)-> BB94,*      ( cond )                     i LIR IBC label gcsafe
BB93  [0310]  1       BB92                  0.32     4 [000..000)-> BB95        (always)                     i LIR IBC internal
BB94  [0311]  1       BB92                  0.41     4 [000..000)-> *           (always)                     i LIR IBC internal label idxlen nullcheck q
BB95  [0315]  2       BB93,BB94             0.73     8 [126..126)-> *           (always)                     i LIR IBC internal label q
BB96  [0011]  2       BB91,BB95             0.73     8 [126..127)-> BB266,*     ( cond )                     i LIR IBC label
BB97  [0318]  1       BB96                  0.73     8 [126..138)-> BB101,*     ( cond )                     i LIR IBC hascall gcsafe
BB98  [0328]  1       BB97                  0.73     8 [131..132)-> BB99,*      ( cond )                     i LIR IBC
BB100 [0330]  1       BB98                  0.00     0 [131..132)-> *           (always)                     i LIR IBC hascall gcsafe q
BB101 [0331]  3       BB97,BB99,BB100       0.73     8 [132..140)-> BB103,*     ( cond )                     i LIR IBC label
BB102 [0012]  1       BB101                 0        0 [000..155)-> BB107       (always)                     i LIR IBC rare q
BB99  [0329]  1       BB98                  0.73     8 [131..132)-> BB101       (always)                     i LIR IBC label
BB250 [0720]  1       BB248               119.27  1312 [19B..19C)-> BB254,*     ( cond )                     i LIR IBC label bwd
BB251 [0725]  1       BB250               119.25  1312 [19B..19C)-> BB252,*     ( cond )                     i LIR IBC bwd
BB253 [0727]  1       BB251                 0.04     0 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB254 [0728]  3       BB250,BB252,BB253   119.27  1312 [19B..19C)-> BB258,*     ( cond )                     i LIR IBC label hascall gcsafe bwd
BB255 [0739]  1       BB254               119.25  1312 [19B..19C)-> BB256,*     ( cond )                     i LIR IBC bwd
BB257 [0741]  1       BB255                 0.04     0 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB258 [0742]  3       BB254,BB256,BB257   119.27  1312 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB259 [0746]  1       BB258               119.27  1312 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB260 [0753]  1       BB259               119.27  1312 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB261 [0717]  2       BB249,BB260         119.27  1312 [19B..19C)-> BB151,*     ( cond )                     i LIR IBC label bwd
BB262 [0452]  1       BB261                 3.73    41 [19B..19C)-> BB282,*     ( cond )                     i LIR IBC bwd
BB263 [0454]  2       BB262,BB281           0.73     8 [19B..1CF)-> BB268,*     ( cond )                     i LIR IBC label hascall gcsafe
BB264 [0017]  1       BB263                 0.55     6 [1CF..???)-> BB265,*     ( cond )                     i LIR IBC hascall gcsafe q
BB267 [0040]  1       BB264                 0        0 [???..???)-> *           (always)                     i LIR IBC rare internal hascall gcsafe q
BB268 [0038]  3       BB263,BB265,BB267     0.73     8 [1DC..1E0)-> BB272,*     ( cond )                     i LIR IBC label
BB269 [0019]  1       BB268                 0.73     8 [1E0..???)-> BB270,*     ( cond )                     i LIR IBC hascall gcsafe q
BB271 [0043]  1       BB269                 0        0 [???..???)-> *           (always)                     i LIR IBC rare internal hascall gcsafe q
BB272 [0041]  3       BB268,BB270,BB271     0.73     8 [1ED..1F1)-> BB67,*      ( cond )                     i LIR IBC label
BB273 [0021]  1       BB272                 0.73     8 [1F1..???)-> BB275,*     ( cond )                     i LIR IBC hascall gcsafe q
BB274 [0045]  1       BB273                 0.73     8 [???..???)-> BB67        (always)                     i LIR IBC internal hascall gcsafe q
BB282 [0822]  1       BB262                 1.86    20 [???..???)-> BB149       (always)                     LIR IBC internal label bwd
BB256 [0740]  1       BB255               119.21  1311 [19B..19C)-> BB258       (always)                     i LIR IBC label bwd
BB252 [0726]  1       BB251               119.21  1311 [19B..19C)-> BB254       (always)                     i LIR IBC label bwd
BB170 [0520]  1       BB172                25.84   284 [19B..19C)-> BB171       (always)                     i LIR IBC label bwd bwd-target align q
BB270 [0042]  1       BB269                 0.73     8 [???..???)-> BB272       (always)                     i LIR IBC internal label hascall gcsafe q
BB167 [0509]  1       BB165                 0.02     0 [19B..19C)-> BB168       (always)                     i LIR IBC label hascall gcsafe bwd q
BB265 [0039]  1       BB264                 0.55     6 [???..???)-> BB268       (always)                     i LIR IBC internal label hascall gcsafe q
BB163 [0498]  1       BB161                 0.00     0 [19B..19C)-> BB164       (always)                     i LIR IBC label hascall gcsafe bwd q
BB103 [0013]  1       BB101                 0.73     8 [000..000)-> BB105,*     ( cond )                     i LIR IBC label gcsafe
BB104 [0343]  1       BB103                 0.32     4 [000..000)-> BB106       (always)                     i LIR IBC internal
BB105 [0344]  1       BB103                 0.41     4 [000..000)-> *           (always)                     i LIR IBC internal label idxlen nullcheck q
BB106 [0348]  2       BB104,BB105           0.73     8 [16A..16A)-> *           (always)                     i LIR IBC internal label q
BB107 [0014]  2       BB102,BB106           0.73     8 [16A..16B)-> BB266,*     ( cond )                     i LIR IBC label
BB108 [0351]  1       BB107                 0.73     8 [16A..17C)-> BB283,*     ( cond )                     i LIR IBC hascall gcsafe
BB109 [0361]  1       BB108                 0.73     8 [175..176)-> BB110,*     ( cond )                     i LIR IBC
BB111 [0363]  1       BB109                 0.00     0 [175..176)-> *           (always)                     i LIR IBC hascall gcsafe q
BB112 [0364]  3       BB110,BB111,BB283     0.73     8 [176..17D)-> *           (always)                     i LIR IBC label hascall gcsafe preheader q
BB114 [0371]  2       BB112,BB113           1.03    11 [17C..17D)-> BB116,*     ( cond )                     i LIR IBC label bwd
BB115 [0372]  1       BB114                 1.05    12 [17C..17D)-> BB113,*     ( cond )                     i LIR IBC bwd bwd-src
BB116 [0373]  2       BB114,BB115           0.73     8 [17C..17D)-> BB266,*     ( cond )                     i LIR IBC label
BB117 [0377]  1       BB116                 0.73     8 [17C..18F)-> BB148,*     ( cond )                     i LIR IBC hascall gcsafe
BB118 [0015]  1       BB117                 0.73     8 [000..000)-> BB120,*     ( cond )                     i LIR IBC idxlen nullcheck
BB119 [0386]  1       BB118                 0        0 [000..000)-> *           (always)                     i LIR IBC rare internal hascall gcsafe q
BB120 [0387]  2       BB118,BB119           0.73     8 [000..000)-> BB139,*     ( cond )                     i LIR IBC internal label idxlen
BB121 [0388]  1       BB120                 0.73     8 [000..000)-> BB123,*     ( cond )                     i LIR IBC internal hascall gcsafe idxlen
BB122 [0391]  1       BB121                 0        0 [000..000)               (throw )                     i LIR IBC rare internal hascall gcsafe newobj
BB283 [0823]  1       BB108                 0.36     4 [???..???)-> BB112       (always)                     LIR IBC internal label
BB110 [0362]  1       BB109                 0.73     8 [175..176)-> BB112       (always)                     i LIR IBC label
BB159 [0465]  1       BB153                 0.56     6 [19B..19C)-> BB266,BB160 ( cond )                     i LIR IBC label bwd
BB113 [0370]  1       BB115                 0.31     3 [17C..17D)-> BB114       (always)                     i LIR IBC label bwd bwd-target q
BB160 [0486]  1       BB159                 0.56     6 [19B..19C)-> BB266,BB161 ( cond )                     i LIR IBC label hascall gcsafe bwd
BB123 [0392]  1       BB121                 0.73     8 [000..000)-> BB284,*     ( cond )                     i LIR IBC internal label idxlen
BB124 [0393]  1       BB123                 0.01     0 [000..000)-> BB125,*     ( cond )                     i LIR IBC internal idxlen
BB285 [0825]  1       BB124                 0.00     0 [???..???)-> *           (always)                     LIR IBC internal q
BB126 [0395]  2       BB125,BB285           0.01     0 [000..000)-> BB127,*     ( cond )                     i LIR IBC internal label hascall gcsafe idxlen
BB136 [0426]  2       BB126,BB135           0        0 [000..000)-> BB138       (always)                     i LIR IBC rare internal label q
BB284 [0824]  1       BB123                 0.36     4 [???..???)-> BB139       (always)                     LIR IBC internal label
BB127 [0804]  2       BB126,BB135           0.01     0 [000..000)-> BB132,*     ( cond )                     LIR IBC internal label idxlen nullcheck bwd
BB129 [0433]  1       BB127                 9.09       [000..000)-> BB286,*     ( cond )                     i LIR internal bwd
BB130 [0434]  1       BB129                 9.09       [000..000)-> *           (always)                     i LIR internal bwd q
BB131 [0435]  2       BB130,BB286           9.09       [000..000)-> *           (always)                     i LIR internal label nullcheck bwd q
BB132 [0436]  2       BB127,BB131           0.01     0 [000..000)-> BB133,*     ( cond )                     i LIR IBC internal label hascall idxlen nullcheck bwd
BB137 [0421]  1       BB132                 0.01     0 [000..000)-> *           (always)                     i LIR IBC internal q
BB138 [0427]  2       BB136,BB137           0.01     0 [000..000)-> *           (always)                     i LIR IBC internal label q
BB139 [0396]  3       BB120,BB138,BB284     0.73     8 [000..000)-> BB140,BB148 ( cond )                     i LIR IBC internal label
BB286 [0826]  1       BB129                 4.55       [???..???)-> BB131       (always)                     LIR internal label bwd
BB125 [0394]  1       BB124                 0.00     0 [000..000)-> BB126       (always)                     i LIR IBC internal label hascall gcsafe q
BB161 [0493]  1       BB160                 0.56     6 [19B..19C)-> BB163,*     ( cond )                     i LIR IBC label hascall gcsafe bwd
BB162 [0497]  1       BB161                 0.56     6 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd q
BB164 [0502]  4       BB157,BB158,BB162,BB163  61.27   674 [19B..19C)-> BB168,*     ( cond )                     i LIR IBC label bwd
BB165 [0507]  1       BB164                61.26   674 [19B..19C)-> BB167,*     ( cond )                     i LIR IBC bwd
BB166 [0508]  1       BB165                61.24   674 [19B..19C)-> *           (always)                     i LIR IBC bwd q
BB168 [0510]  3       BB164,BB166,BB167    61.27   674 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB169 [0514]  1       BB168                61.27   674 [19B..19C)-> *           (always)                     i LIR IBC hascall gcsafe bwd preheader q
BB171 [0521]  2       BB169,BB170          87.12   958 [19B..19C)-> BB173,*     ( cond )                     i LIR IBC label bwd
BB172 [0522]  1       BB171                88.44   973 [19B..19C)-> BB170,*     ( cond )                     i LIR IBC bwd bwd-src
BB173 [0523]  2       BB171,BB172          61.27   674 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB174 [0527]  1       BB173                61.27   674 [19B..19C)-> BB176,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB175 [0531]  1       BB174                 9.09       [19B..19C)-> BB236       (always)                     i LIR bwd
BB152 [0449]  1       BB151                61.27   674 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC label bwd
BB153 [0458]  1       BB152                61.27   674 [19B..19C)-> BB159,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB154 [0464]  1       BB153                60.71   668 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC bwd
BB155 [0470]  1       BB154                60.71   668 [19B..19C)-> BB266,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB156 [0477]  1       BB155                60.71   668 [19B..19C)-> BB158,*     ( cond )                     i LIR IBC hascall gcsafe bwd
BB157 [0481]  1       BB156                60.46   665 [19B..19C)-> BB164       (always)                     i LIR IBC hascall gcsafe bwd
BB158 [0482]  1       BB156                60.71   668 [19B..19C)-> BB164       (always)                     i LIR IBC label hascall gcsafe bwd q
BB275 [0046]  1       BB273                 0.73     8 [1FE..1FF)-> BB67        (always)                     i LIR IBC label hascall gcsafe q
BB276 [0816]  0                             0          [???..???)               (throw )                     i LIR rare keep internal label
BB197 [0597]  1       BB199                 0        0 [19B..19C)-> BB195       (always)                     i LIR IBC rare label bwd bwd-target q
BB211 [0620]  1       BB213                 0        0 [19B..19C)-> BB209       (always)                     i LIR IBC rare label bwd bwd-target q
BB231 [0656]  1       BB229                 0        0 [19B..19C)-> BB235,*     ( cond )                     i LIR IBC rare label hascall gcsafe bwd
BB232 [0666]  1       BB231                 0        0 [19B..19C)-> BB234,*     ( cond )                     i LIR IBC rare bwd
BB233 [0667]  1       BB232                 0        0 [19B..19C)-> BB235       (always)                     i LIR IBC rare bwd
BB234 [0668]  1       BB232                 0        0 [19B..19C)-> *           (always)                     i LIR IBC rare label hascall gcsafe bwd q
BB235 [0669]  3       BB231,BB233,BB234     0        0 [19B..19C)-> BB236       (always)                     i LIR IBC rare label bwd q
BB133 [0422]  1       BB132                 0        0 [000..000)-> BB135,*     ( cond )                     i LIR IBC rare internal label idxlen bwd
BB134 [0423]  1       BB133                 0        0 [000..000)-> *           (always)                     i LIR IBC rare internal bwd q
BB135 [0424]  2       BB133,BB134           0        0 [000..000)-> BB127,BB136 ( cond )                     i LIR IBC rare internal label idxlen bwd
BB140 [0397]  1       BB139                 0        0 [000..000)-> BB148,*     ( cond )                     i LIR IBC rare internal label idxlen
BB141 [0398]  1       BB140                 0        0 [000..000)-> BB148,*     ( cond )                     i LIR IBC rare internal hascall gcsafe idxlen nullcheck
BB142 [0399]  1       BB141                 0        0 [000..000)-> BB144,*     ( cond )                     i LIR IBC rare internal hascall idxlen nullcheck
BB143 [0400]  1       BB142                 0        0 [000..000)-> BB145       (always)                     i LIR IBC rare internal
BB144 [0401]  1       BB142                 0        0 [000..000)-> BB146       (always)                     i LIR IBC rare internal label q
BB145 [0403]  1       BB143                 0        0 [000..000)-> BB147       (always)                     i LIR IBC rare internal label
BB146 [0404]  1       BB144                 0        0 [000..000)-> *           (always)                     i LIR IBC rare internal label q
BB147 [0405]  2       BB145,BB146           0        0 [000..000)-> BB148       (always)                     i LIR IBC rare internal label hascall gcsafe nullcheck
----------------------------------------------------------------------------------------------------------------------------------------------
```

Note that there are some `cond` blocks that don't have a false fall-through, e.g.,
```
BB135 [0424]  2       BB133,BB134           0        0 [000..000)-> BB127,BB136 ( cond )                     i LIR IBC rare internal label idxlen bwd
```

Note that what previously would have been `BBJ_NONE` is now `-> *           (always)`. We could print even less in this case by (unambiguously) printing none of this text, but this minimal output seems like a good start.

BruceForstall commented:
@amanasifkhalid @dotnet/jit-contrib PTAL

BruceForstall committed:
Fix clang build breaks

BruceForstall commented:
[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=536799&view=ms.vss-build-web.run-extensions-tab) (of course)

jakobbotsch commented:
> 2\. If a branch goes to the next block that will be displayed in the table, display "*" instead of the block number, as a shorthand to reduce visual clutter, e.g., `BB14,*      ( cond )`

Personally I think I would prefer to always see the block numbers, instead of having to go make the mental note of what the next lexical block is in these cases.

jakobbotsch commented on src/coreclr/jit/fgdiagnostic.cpp:
```diff
@@ -1924,51 +1927,87 @@ void Compiler::fgTableDispBasicBlock(BasicBlock* block, int ibcColWidth /* = 0 *
-                       maxBlockNumWidth - max(CountDigits(block->GetTarget()->bbNum), 2), "", label);
+                label             = (flags & BBF_KEEP_BBJ_ALWAYS) ? "ALWAYS" : "always";
+                printedBlockWidth = 3 /* "-> " */ + 9 /* kind */;
+                printf("-> %s", dspBlockNum(block->GetTarget()));
```
Now that you're here, can you make it work for `!HasInitializedTarget()`? Currently dumping blocks while they aren't fully initialized can hit assert inside `GetTarget()`.

BruceForstall replied:
So in particular, when bbTarget is nullptr?

amanasifkhalid replied:
Yes

amanasifkhalid commented on src/coreclr/jit/block.cpp:
```diff
@@ -1856,10 +1856,10 @@ weight_t BasicBlock::getCalledCount(Compiler* comp)
+//    one call to the method.
 //
-weight_t BasicBlock::getBBWeight(Compiler* comp)
+weight_t BasicBlock::getBBWeight(Compiler* comp) const
```
Thanks for the const cleanup. Quick question for my own learning: How liberally should I interpret the [guidelines](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#1518-const) on const usage? Should I use it for short-lived locals (ex: `const bool removableJump = block->JumpsToNext() && ...`) in future changes?

BruceForstall replied:
Using `const` on member functions is particularly useful, as the linked guidlines say. In addition, it does make sense to use it as frequently as possible in new code. For short-lived locals, it's a useful "documentation", especially in large functions, where a reader doesn't need to go searching for any modifications to the variable. Some cases like `const Type * const p;` sometimes seems too verbose, but if it helps understanding, go for it!

amanasifkhalid reviewed (approved):
Pending the `!HasInitializedTarget()` fix, LGTM. Thanks for doing this!

For `elideNext`, I like the convenience of being able to, for example, eyeball a `BBJ_ALWAYS` block and tell it's functionally equivalent to `BBJ_NONE` without looking at the next block. But if we decide to disable `elideNext` by default, I don't see myself bothering to turn it on locally. So if general sentiment is against it, I say we remove it.

BruceForstall committed:
Introduce DOTNET_JitDumpTerseNextBlock for new "lexical next" behavior.

Allow block dump to work with null (uninitialized) block targets.

BruceForstall commented:
I introduced `DOTNET_JitDumpTerseNextBlock` to control the behavior of displaying lexical next block in a more terse manner (namely, as `*` instead of the explicit block number). Set to `1` to enable the new, terse behavior. Default is `0` (current behavior).

Also, added code to allow dump of null block targets.
